### PR TITLE
Untrivializing AI upload boards/Ending the upload war meme

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

You know those rounds where we get stuff like
AI law changes:40
And we all laugh and feel bad for the AI?
Yea well we should address that

Divina/Nist explains the issue better than I:
https://docs.google.com/document/d/1NVwuxMHPRKJVogR2P1l8a8LgrdUL-J3on-tGSHAu3GU/edit?usp=sharing

I'm open to rebalancing the specific cost for the board but it's set to 7500 for now

If you feel this nerfs traitors too much do propose alternatives please
## Why It's Good For The Game

Laws are gooder, spam is less
## Changelog
:cl:
add: AI uploads may be ordered in cargo for 7500 credits
balance: AI uploads may no longer be researched or printed
/:cl:

### CODE REVIEW

**Comment by necromanceranne** [comment]
Intense. Only one board to game now.

**Comment by Mycah142** [comment]
Another solution on top of Remove Printability: make them only orderable through a CentCom communication (which puts more work on admins), or a high TC uplink cost, along side the hacked module.

**Comment by Tlaltecuhtli** [comment]
ai upload printability means subverted ais wont get insta killed instead kept in round, if you happen to be blind and unable to look at a gps its your fault

**Comment by Mycah142** [comment]
Or it encourages actually retrieving the core, carding it, and doing an AI Integrity Restore. Which is a thing I've never seen used.

**Comment by TiviPlus** [comment]
I have specifically not given traitors an easier time, as while it _is_ more difficult/engaging to subvert, since you actually have to steal it now, the crew can also not walk up to RnD to undo everything that you did

**Comment by Mycah142** [comment]
But how am I supposed to let the silicons antag for me if it's made difficult?

File: code/modules/cargo/packs.dm [hunk 1]
```diff
@@ -1430,6 +1430,15 @@
 	crate_name = "robotics assembly crate"
 	crate_type = /obj/structure/closet/crate/secure/science
 
+/datum/supply_pack/science/aiupload
+	name = "AI Upload Board"
+	desc = "AI Upload Board to change the AIs laws. Garaunteed to purge 99% of fleshbags with just one swipe! Requires AI upload access to open."
```
Comment (Fikou) [inline code/modules/cargo/packs.dm:1435]:
```suggestion
	desc = "AI Upload Board to change the AIs laws. Guaranteed to purge 99% of fleshbags with just one swipe! Requires AI upload access to open."
```

**Comment by LilPenpusher** [comment]
While I agree that the "Haha, Law 40: Stinky poopoo!" situations are idiotic and require fixing, I'm not sure if this specific change will resolve that specific problem. 

All you're really doing is give more power to the Quartermaster and Cargo Techs by enabling them to literally just order AI Upload boards (why Centcom is putting these things on the market for sale for us to order is beyond me). Essentially this change will make EVERYONE but Cargo weaker, including R&D which is kind of in a weird place already. 

Whether "give power to Cargo, remove power from everyone else" is a good thing or not is up to your own interpretation. Prepare for a lot more independent Cargonia rounds where the QM subverts the AI with an Upload in his Office, though.

**Comment by WarlockD** [comment]
Just a thought but maybe have any upload console not in the AI upload chamber or station be sent a message to the com consoles with its gps position, like its done when the shuttle is called?  At Least you could tell if its happening in maintenance or the clowns bedroom.

**Comment by ATH1909** [comment]
Perhaps the syndicate briefcase full of cash will finally rise again...

**Comment by ATH1909** [comment]
I actually haven't seen too many upload wars recently, and I say that as a silicon main.

Then again, maybe I just play on different servers than you do.

**Comment by kriskog** [comment]
Just make the law change boards single use. Heck, we might even see some non-standard lawsets used out of necessity.

**Comment by uomo91** [comment]
AI upload spam is already against the rules and even things far short of literal spam are frequently subject to admins making the AI malf, and competency of either traitors or the crew negates this entire "problem" from the get-go. It's not _that hard_ to give the AI properly worded stealthy antag laws if you don't want to risk them getting prematurely purged/reset, and it's not _that hard_ for sec/command/R&D/validhunters to track down clandestine uploads, weather they were made by murderous traitors or the crew. As Tlaltecuhtli said:
>if you happen to be blind and unable to look at a gps its your fault

I disagree with:
>This change creates the most interactive way for a traitor to subvert the AI (Claim the board by Secure Tech Storage or...

I guess I'm not seeing how hacking/emagging two doors to get into secure storage is way more interactive than hacking/emagging one door to get into R&D or robo. If anything, as somebody who subverts silicons a lot, to my best recollection I almost always am able to attract less attention by taking the board from tech storage than from a lathe somewhere.

File: code/modules/cargo/packs.dm [hunk 2]
```diff
@@ -1430,6 +1430,15 @@
 	crate_name = "robotics assembly crate"
 	crate_type = /obj/structure/closet/crate/secure/science
 
+/datum/supply_pack/science/aiupload
+	name = "AI Upload Board"
```
Comment (Fikou) [inline code/modules/cargo/packs.dm:1434]:
```suggestion
	name = "AI Upload board crate"
```

**Comment by Timberpoes** [comment]
I would personally be against this solution. You're removing something that should be unique to science and giving it to... Cargo? Why should cargo get the power to generate AI upload boards?

I always felt the problem was less science and more that departmental circuit printers actually limit what you can make. Here's a hot take for you. Engineering can print every circuit. Including AI uploads. Including AI law boards. Because it has the non departmental circuit imprinter that has absolutely no restrictions.

Similarly, limiting upload board availability is a bit crazy. Break into secure storage to get the spare. Order the AI to break their normal upload. Space that board. Now the only way to change AI laws is rubbing the core itself and the antag still has the spare.

I'd oppose this in favour of any combination of two solutions. Add a GPS signal to upload boards and/or limit their creation to only science departmental circuit printers so you know exactly where they came from.

Or just keep it how it is. That's part of silicon weaknesses and law wars occasionally happen.

**Comment by HomocidlePancakes** [comment]
This is a good change, but I’ve noticed a flaw. Originally, I though “Hey, RDs will now secure the board roundstart”, but then I realized that there is nothing stopping the AI from just hiding it at sat. 

I suggest you put the AI upload in to RDs locker to the prevent **EPIC GAMERS**™️, be it organic or silicon, rushing it before anyone notices.

**Comment by optimumtact** [comment]
kriskog is on to something

**Comment by LilPenpusher** [comment]
> Just make the law change boards single use. Heck, we might even see some non-standard lawsets used out of necessity.

This. 100% this. I'd go a step further and make the law modules in the AI Upload reusable, but non-printable (or admin spawned), while the crew themselves can only print the one-use law modules. This means the Crew/Command always have the failsafe of the Upload unless a subverted AI blows the Upload Chamber up before people notice and spaces the boards. On the other hand, it also means people can still subvert the AI like they can now (since a good Antag doesn't need more than one custom law anywho) and people who want to add flavour laws can also still do so (like making Lizards or Moths human to the AI, which people tend to want to do). 

This suggestion really would solve everything, in my opinion. Stops endless law wars which make the AI want to shoot their brains out, but keeps modules and upload boards at current availability so that people can still interact with the AI laws if they really do want to.

**Comment by TiviPlus** [comment]
> I actually haven't seen too many upload wars recently, and I say that as a silicon main.

You said you were were a moth main on the lamp PR _???_ You also have under 10% of your hours as silicon so idk what you're getting at here. I made this PR after me and two other AI mains talked about it, I dont see what you are trying to accomplish here with your whining.

>Just make the law change boards single use. Heck, we might even see some non-standard lawsets used out of necessity.

Sounds neat, but then you nerf freeform heavily, limiting creative laws
But if we make em multiuse it makes it more annoying so ehhh I need to sleep on how to do this exactly but soundsgood to me, I'd be down for this

>QM rant

As nonantag you still need to get the law boards, and 7500 credits and a ID secure crate dont magically get dispensed out of printers

>if you happen to be blind and unable to look at a gps its your fault

Fun tip: heres a trick I've abused and seen abused to go around the GPS: 
literally just screwdriver the console
Thats it
It will never be found

>I guess I'm not seeing how hacking/emagging two doors to get into secure storage is way more interactive than hacking/emagging one door to get into R&D or robo. If anything, as somebody who subverts silicons a lot, to my best recollection I almost always am able to attract less attention by taking the board from tech storage than from a lathe somewhere.

Lets put it like this: breaking into sci entails going: "Ai open", while breaking into tech storage leaves noticable prints. Making it not printable also ensures scientists dont take 30 seconds to undo subversions

>Timberpoes essay

Yes we could just add a GPS signal, but that just nerfs illicit law changes and people that build uploads to unsubvert 30 seconds after the AI does something bad will be in no way hampered. Limiting it to the Sci printer also does nothing in this regard.

>Pancakes

Mhmm I dunno honestly, Tech storage also holds the only spare comms board and I dont see people stealing that too often to call/recall. RDs locker already has a tot objective and important console nearby too so idk. I suggest we perhaps see how that plays out first.

Also random fun fact I while writing this I had to stunarm someone and take their upload board because they were chain-uploading law 4: poggers..... and law 5:a

**Comment by Nistishere** [comment]
So, some bullet points to adress concerns from discussing with tivi. 
-While ai upload boards are bought from cargo, they seem to need ai upload access, so noone would naturally have access to the box in the cargo department. If they really, really wanted to subvert the ai, maybe adding something like a ping on one of the ai's consoles in their sat + command consoles saying "ai board purchased in cargo at X time" could work. I think it also should be a bit more pricey, 10k+, but that's a "my opinion" deal.

-moving upload board from secure tech to rd's office is a very good idea. Hide it in that purple locker. Effectively, its the same as keeping it where it is. Personally, I feel it should be punishable for the silicones if they try to tide boards without reason, but that's discussing rules not code.

-adding gps signals to boards : Not that many people even bother to chase after uploads in the first place as it is right now, and active uploads can be kept operational without people finding out uenless someone is simpathetic enough to go after those signals whenever they pop up, looking for antags to valid.

-one use boards: another good idea, but resources to print law boards are an absolute non-issue the moment t4 orm is upgraded and one miner drops their haul.

**Comment by LilPenpusher** [comment]
> Sounds neat, but then you nerf freeform heavily, limiting creative laws
> But if we make em multiuse it makes it more annoying so ehhh I need to sleep on how to do this exactly but soundsgood to me, I'd be down for this

Make the Law Boards in the Upload limitless use, like right now, but non-printable. The Crew can print one-use boards at Science (or Engineering, technically). 

Personally, if you have a fun custom lawset you could just ask the Captain or RD to add it for you. If anything, this might encourage people to interact more with the Heads of Staff (short of when the Heads march in and Baton their asses). Unless a Clown really wants to spend all the resources in the ORM for 3 custom laws, which will ENSURE he gets lynched by basically any job that uses the ORM for resources.

**Comment by Timberpoes** [comment]
> 
> 
> Just make the law change boards single use. Heck, we might even see some non-standard lawsets used out of necessity.

> 
> 
> kriskog is on to something

How would this concept interact with traitor hacked law upload boards? Could a traitor now keep subverting the AI until the station just runs out of materials to print new boards or would THAT also be single-use with a TC cost change?

Would this solve the underlying issue, or would it just change it from upload wars to circuit printer wars? If you can access a circuit printer to make law boards, you can also print out your own circiuit imprinter, dump materials from the protolathe next to it (so now nobody can print additional law boards until the miners return with their new ore haul) and just build your own AI law board factory. You win the AI upload war.

Do we want to encourage the RD and Captain carrying AI law boards around 24/7 because you may not even be able to de-subvert the AI from the upload chamber once you've used the first round of Reset and/or Purge boards?

## On the topic of this particular PR

I would like to put out a reminder that the main reason we are able to prevent AI validhunting when it's subverted is because breaking into the upload to de-subvert it is usually equally as easy or easier than breaking into the core to card or kill it. On maps where the AI upload isn't on the AI sat, this is even more important. When you remove the option of resetting the AI's laws, you start to approach round removal as the primary method of dealing with subverted AIs.

Carding the AI is not an option. You can't change the AI's laws on the card. You still need an upload or a core for the AI. If the AI is subverted and you can't change its laws, it's probably being round removed until you can. If the only other AI upload is on a traitor... That AI is not leaving that card.

One of the beautiful things with the AI laws being easy to change is that de-subverting the AI is as easy as subverting it - Oftentimes people can do it even easier than going to the upload. Cut the cam in engineering, RnD or robotics, build an AI upload and law boards, fuckin' fix it. Then be on the alert for illegal uploads and monitor a GPS at all times. The AI still gets to play its round.

The harder you make laws the change, the more attractive "remove the AI from the round entirely" becomes as a solution to a subverted AI. As RD I've regularly seen AIs resist law changes by destroying uploads, destroying circuit imprinters, hiding law boards, so on, so forth. It usually ends with the AI round removed.

>
>
>Make the Law Boards in the Upload limitless use, like right now, but non-printable. The Crew can print one-use boards at Science (or Engineering, technically).

Same again. If the AI gets subverted with a correct law, you can literally just ask it to send you a borg, go straight into their upload and take the only multi-use boards on the station. Or ask the borgs to take in a crate and bring you their law boards. Borgs are always dragging shit throughout the station. Delivering to cargo. Easy excuse. De-subverting the AI in this scenario is now more annoying than just round removing it.

AI players need to think long and hard about this decision. I play science/RD. If the AIs want the primary solution for managing bad lawsets to be carding and/or round removal, then fine. But changes to this system disturb a delicate balance - The easier the AI laws are to change, the easier it is to justify changing them back to manage a subversion compared to round removal. The harder it is to change AI laws, the more attractive "crash and trash the AI" looks as an option.

**Comment by Cartographer-D** [comment]
In addition to this I would also

Make uploading new laws have a recharge timer while the AI compiles the new law. (1-5 mins, whatever makes sense)
Limit the number of uploads to like, 3 per upload console.

I also like the idea of limiting the number of board uses.

With this change it definitely isn't as easy for SOME people to get ahold of an upload, but easier for others (basically enabling cargo to do it instead of research).

**Comment by zxaber** [comment]
I play AI quite a bit and upload wars are pretty rare now (especially ever since the cost increase).

Anyway, I'm not really sure how giving the boards to Cargo helps at all. This is pretty much asking for a Cargo subversion every other round.

**Comment by TheMidnightRose** [comment]
My addition on this when we were talking about it with a couple of others was to move BORIS/AI Mech Controllers into a separate technode with the Techweb points for AI tech being raised to 10k Research points, which would put it in line with Bluespace Parts and all the Weapon techs. Right now, AI tech is a measly 2500 research points.

 I do agree with Timber with the fact that Engineering being able to print whats supposed to be RESEARCH ONLY boards a bit annoying. Departmentalize them like the techfabs.

**Comment by TiviPlus** [comment]
Okay I've slept on it and heres what I have:

Upload boards will get a GPS signal

For a second PR I'll do after this since this requires actual coding:
Make all core laws single-use, except the roundstart non-core freeform
Make all printable boards single-use and more expensive

**Comment by zxaber** [comment]
I don't really think a GPS signal on the board is really necessary; it's already on the upload console, and the console was given a longer disassemble time to make it easier to track down someone hiding in maint and repeatedly subverting an AI.

It's better than moving the boards to Cargo, though.

**Comment by Timberpoes** [comment]
I still think this whole PR and concept is a solution desperately looking for a problem to apply to it and I would resist any PR to change it.

I 110% think this deserves discussion elsewhere and gathering of silicon mains thoughts before any implementation. I say this as an RD/science player who would love to have my life made easier when it comes to managing AI laws and subverted AIs.

People fighting for control over the AI is a feature. It's not a meme, bug or glitch. Sometimes the AI gets caught in the crossfire. This makes the AI role the same as any other player, their core may sit in their ivory castle however they're still just as vulnerable as anyone else.

Unless sillycone mains think it's a problem, this is one piece of code I'd see left well enough alone.

**Comment by TiviPlus** [comment]
This is not a made-up problem as otherwise this PR wouldnt exist at all.

>I say this as an RD/science player who would love to have my life made easier when it comes to managing AI laws and subverted AIs.

This also goes both ways, as you yourself said:
Capture
The intent is precisely to NOT make AI management a matter of walking into RnD but an active action you work for.

As you yourself said "[...] the things the AI (and their cyborg army) can do are very, very, VERY strong. Now crew members have to walk around snipping every camera they see. Any door can be electrified. Any room vented. Any room flooded with plasma. You slip up and before you know it, you're in a tritium fire." 
Control over this should not by any means be easy; power is a privilege, not a right.
In the current state of affairs it is entirely too easy to upload laws, meaning that anyone can snag a board in passing and go do things like this:
Capture
Capture1

Then someone decides that they dont like how someone uploaded laws and go ahead and go for the easy way out of making their own upload and removing the laws, after which they are re-uploaded, Rinse/repeat: upload war.

**Comment by zxaber** [comment]
I've been on all side of this scenario quite often. Sometimes the AI gets subverted and they rampage. Sometimes someone grinds uranium and EMPs the AI to death. Sometimes you make your own upload and fix the AI and then camp a GPS until the subverter's upload shows up and you roll in and flatten him.

Sometimes people law spam and you just have to deal with it. It's a rare occurance and even as an AI player, I'll take the occasional law spam in return for rounds where people get an upload and make funny harmless laws. I argue Asimov is a good starting lawset because it's inherently flawed, but that argument is with the context that law changes are eventually available to people besides the Cap and RD.

I suppose the question really is "What do you want out of these PRs?" A GPS signal on the board means that subverting the AI will be essentially not viable with a printed board. Even when you disassemble the upload, anyone can check a GPS, see that the upload existed at some point in the past, and throw out any stealth you had. Is your goal to cut subversions to be basically never?

**Comment by TheMidnightRose** [comment]
We could also swing another direction and parse out the AI law boards over a couple of technodes with each being 10k Tech point requirments. With AI Control being the starting one, which would give Reset, Boris, AI Mech Beacon, And Asimov. The two law boards being one use. Spread the other law boards over their own nodes nodes based on severity. Like Nanites.

**Comment by TiviPlus** [comment]
Back to roots then,
The core of the issue is how EASY it is to get the boards for literally everyone. It currently takes 0 effort to get to a board, and anyone can wander up to the lathe to print themselves an upload.
Then the symptoms of this are meme laws/law wars and etc
Slapping boards later on the tree will just make the same behavior happen later in the round to little actual gain. Resources as soon as T4 is hit are effectively a nonissue so gating them behind that is very ehh, which is why I gated it behind an ID requirement as well as cargo, where while yes, it is easy to get cash, it is very easy to see who got what when and requires additionnal effort to open it.

Does this impede funny laws? Not really, since you can ask the (acting) Cap or RD to upload them for you or if you're the clown then break into secure tech storage and deal with the consequences.

**Comment by uomo91** [comment]
> > if you happen to be blind and unable to look at a gps its your fault
> 
> Fun tip: heres a trick I've abused and seen abused to go around the GPS:
> literally just screwdriver the console
> Thats it
> It will never be found

It can pretty trivially be found and it's happened to me before.
The stated problem this PR was solving apparently was upload wars/spam, and if somebody is continually rebuilding an upload to add their laws again after purge, literally just throw a GPS at the warden or RD, tell them to watch for the signal popping up (you can keep the window open and out of the way just like the handheld crew monitoring window). Relay the coords, to which a sec officer or general crew member can rush over and probably find the person nearby still or at least find the empty/deactivated console and check it and the room for prints.

> > I guess I'm not seeing how hacking/emagging two doors to get into secure storage is way more interactive than hacking/emagging one door to get into R&D or robo. If anything, as somebody who subverts silicons a lot, to my best recollection I almost always am able to attract less attention by taking the board from tech storage than from a lathe somewhere.
> 
> Lets put it like this: breaking into sci entails going: "Ai open", while breaking into tech storage leaves noticable prints. Making it not printable also ensures scientists dont take 30 seconds to undo subversions

If somebody is law 2'ing their way into R&D to print an upload board they're not very good at subverting the AI and they're going to get ratted out as soon as it gets purged if not the second the AI sees them with it.

I still don't see why silicon law changes are some enormous problem that needs a rework making it nearly impossible for anyone to do, but if you feel that way, why not make upload board printing  give the AI a notification or something?

**Comment by Tlaltecuhtli** [comment]
the example you have shown is even flawed a dude made an ai upload in escape and spammed laws, like was no one else alive to see the upload in a public hall?
printing a upload makes a log in ore silo of when and where it was printed, researching laws logs your name in the server room, gps signals pinpoints where you were which you can rig with traps, this is interesting  detective gameplay to find who dunnit, while you want to turn ai upload into plasma flood pipe 2.0, something you just either remove roundstart or just watch on camera the whole round as the only way evil can happen is someone walk into tech storage

making ai rogue is strong yes, if you are dealing with idiots(like in the screen shot you shown, A FUNCKING UPLOAD IN ESCAPE) , as once someone fixes ai you ll also have an "ai says "x subverted me" and an "  4. x is not human "  law uploaded within seconds

also why cargo? why not make it need the clown and chaplain stamp used on an ancient burial site in lavaland? why are you putting something that makes sense in rnd into hands of random people? do you think the idiots that couldnt find a gps or an ai upload in the  hallway are gonna have enough brain matter to order one from cargo?

File: code/modules/cargo/packs.dm (hunk 2 above)
Comment (ATH1909) [reply code/modules/cargo/packs.dm:1434]: er, the capitalization for this suggested change looks a bit weird

**Comment by TiviPlus** [comment]
The example I gave was actually more mild due to the fact that the board was in evac, as it was promptly found and spaced by silicons, which however can not be said for the usual location of uploads: maint. 
Even so it even highlights the central issue here: how trivially we treat changing the intentions of one of the most powerful actors on the station and their borg army.
Making the AI rogue is strong yes, and should depend on skill factor in evading another and not on whoever can print the most uploads to click on.
Again, control of the AI should be a PRIVILEGE and not a RIGHT.
You also compare it to plasma pipes, which is actually a good comparison seeing as plasma pipes require effort to set up and are in a relatively secure area. Applying this example in reverse would be the equivalent of having these pipes in the middle of RnD, where they would constantly be flicked on and off by people trying to be funny/ antags that you wouldn’t even be able to see since they’re halfway across the station.
As to the alternate solution of giving the AI a alert, it does address limiting the upload for antags, but security and sci will have an easy time unsubverting the AI, turning it into an antag nerf instead
It’s also cargo because it allows a replacement board to be acquired should it be needed while not making it too easy to get the board for antags.

**Comment by zxaber** [comment]
Honestly I think you're making a moutain out of a molehill. We have a few existing methods of dealing with uploaders (GPS on the upload console while it's got a longer deconstruct time than other consoles and/or carding the AI while you track them down, for example). If the uploader ever gave out their name to the AI, fixing the AI also means knowing the uploader's identity. And upload wars are very few and far-between, certainly not worth the loss of pretty much the entire mechanic.

The system we have is fine. Sometimes shit happens, and people subvert the AI. It's not like it's a thing every single round, and when crew law changes do happen they're often harmless meme laws. I don't see any real reason to pile on extra restrictions to it.

**Comment by TiviPlus** [comment]
Again, this is not just about upload wars.
Its about AI upload boards being piss-easy to get, making everything concerning them a matter of "Just go to sci and make more"

The point is to flatten the availability of AI boards across the board in order to increase the impact of law changes and to make less "Me no like law so me change so me print and smash boards together until it better" for whoever gives a damn.
Are meme laws like "You must lube the halls" funny? Yes. Are noeffort onehumans and law 4: a funny: no.
If you have put in enough effort to think up something funny to upload you should also be ready to face the consequences as well as put effort into acquiring an upload.

**Comment by zxaber** [comment]
Ironically, as it stands, simply one-humaning the AI is not a fire-and-forget solution, much because of what you wish to kill off (the ease of restoring the AI to crew-friendly terms). If you just subvert the AI and tell it to plasma flood, expect to see your AI no longer on your side very soon. Subverting the AI is only one step, and you'll want to work with the AI and related silicons to fly under the radar while you do your objectives or at the very least get things set up.

It really sounds to me like you want to front-load all the effort of a subversion; as if once the traitor has managed to subvert the AI, they should be guaranteed a loyal AI forever (barring some incompetence on the traitors part). Subvert the AI, order them to be moved to a safe(r) location, and plasma flood immediately.

I don't agree with any part of this PR. To be honest, I don't agree with the linked document you provided by Nist. Yes, sometimes someone getting an upload can lead to a painful round, but very rarely does it get that bad. Uploaded laws can provide entertainment, whether they come from the Captain, or the clown, or a traitor. We have in place gameplay options that work around it, and that they exist mean that your still have to be careful once the silicons are under your control. I'd argue that's a better system than trying to restrict upload access to anyone that isn't Cap, RD, or traitor.

**Comment by IndieanaJones** [comment]
Just make it so if so many uploads happen within a span of time, the AI gets blanked and the upload doesn't work anymore.

**Comment by TheSilverNuke** [comment]
If I may add to this I've thought about this subject before. I've put a fairly significant amount of time into playing as an AI and I feel that this could be resolved by making AI uploads a machine instead. For example, you could make it so that the AI upload and robotics control console are merged into one machine that takes up a 1x3 area and has a GPS signal. This would make it hard to set up and dismantle (or at least not as easy to hide). Since we're on the subject of robotics control consoles I also think that adding a cooldown to detonating borgs would be a good idea as well. 20-30 seconds or so. That's my input anyway.

**Comment by ExcessiveUseOfCobblestone** [comment]
I would be fine with 1use as well as spreading them out over techwebs.

the doc you provided is actually suggesting tomake them non-renewable, not just removing them from printing via science. With this change you are still "printing" them, just with a different set of restrictions to "print".

**Pull request closed without merging**